pkg/faucet/faucet.go

```diff
@@ -0,0 +1,246 @@
+	return func(res http.ResponseWriter, req *http.Request) {
+		// Extract the recipient address from the URL path.
+		denomPath := fmt.Sprintf(denomPathFmt, denom)
+		recipientAddressStr := req.URL.Path[len(denomPath):]
+		recipientAddress, err := cosmostypes.AccAddressFromBech32(recipientAddressStr)
+		if err != nil {
+			logger.Logger.Error().Err(err).Send()
+			res.WriteHeader(http.StatusBadRequest)
+			if _, err = res.Write([]byte(err.Error())); err != nil {
+				logger.Logger.Error().Err(err).Send()
+			}
+			return
+		}
+
+		logger := logger.Logger.With("recipient_address", recipientAddress)
+
+		// Check if the recipient address already exists onchain.
+		shouldSend, queryErr := srv.shouldSendToRecipient(req.Context(), recipientAddress)
+		if queryErr != nil {
+			respondInternalError(logger, res, queryErr)
+			return
+		}
+
+		// If the create_account_only is true, address already exists onchain.
+		// Return a "not modified" response.
+		if !shouldSend {
+			respondNotModified(logger, res, recipientAddress.String())
+			return
+		}
+
+		// If the address doesn't exist onchain, send it tokens.
+		if err = srv.SendDenom(ctx, logger, denom, recipientAddress); err != nil {
+			respondBadRequest(res, err)
+			return
+		}
+
+		// Send empty success response (200).
+		if _, err = res.Write([]byte{}); err != nil {
+			logger.Error().Err(err).Send()
+			return
+		}
+	}
+}
+
+// shouldSendToRecipient indicates whether the faucet should send tokens to the given recipient address.
+// When create_accounts_only is false:
+// - ALWAYS returns true
+// When create_accounts_only is true:
+// - ONLY return true IF the recipient address does NOT already exist onchain (i.e. has no balances)
+func (srv *Server) shouldSendToRecipient(ctx context.Context, recipientAddress cosmostypes.AccAddress) (bool, error) {
+	if srv.config.CreateAccountsOnly {
+		_, err := srv.GetBalances(ctx, recipientAddress.String())
+		switch {
+		case errors.Is(err, query.ErrQueryBalanceNotFound):
+			return true, nil
+		case err != nil:
+			return false, err
+		default:
+			// The account has a balance; therefore, it already exists onchain.
+			return false, nil
+		}
+	}
+
+	// Send by default.
+	return true, nil
+}
+
+// SendDenom sends tokens of the given denomination to the given recipient address.
+// The amount of tokens sent is determined by the faucet's supported_send_coins configuration.
+func (srv *Server) SendDenom(
+	ctx context.Context,
+	logger polylog.Logger,
+	denom string,
+	recipientAddress cosmostypes.AccAddress,
+) error {
+	isDenomSupported, denomCoin := srv.config.GetSupportedSendCoins().Find(denom)
+	if !isDenomSupported {
+		return fmt.Errorf("denom %q not supported", denom)
+	}
+
+	sendMsg := banktypes.NewMsgSend(
+		srv.config.signingAddress,
+		recipientAddress,
+		cosmostypes.NewCoins(denomCoin),
+	)
+
+	txResponse, eitherErr := srv.config.txClient.SignAndBroadcast(ctx, sendMsg)
```

Comment:
Note that `SignAndBroadcast` uses `BroadcastTxSync` which means the response would contain the error in its `log[]` field